vpr/src/place/place.cpp

```diff
@@ -670,30 +793,64 @@ void try_place(const t_placer_opts& placer_opts,
             critical_path = timing_info->least_slack_critical_path();
             sTNS = timing_info->setup_total_negative_slack();
             sWNS = timing_info->setup_worst_negative_slack();
+
+            //see if we should save the current placement solution as a checkpoint
+            if (placer_opts.place_checkpointing && agent_state == 2) {
+                if (placement_checkpoint.cp_is_valid() == false || (timing_info->least_slack_critical_path().delay() < placement_checkpoint.get_cp_cpd() && costs.bb_cost <= placement_checkpoint.get_cp_bb_cost())) {
+                    placement_checkpoint.save_placement(costs, critical_path.delay());
+                    VTR_LOG("Checkpoint saved: bb_costs=%g, TD costs=%g, CPD=%7.3f (ns) \n", costs.bb_cost, costs.timing_cost, 1e9 * critical_path.delay());
+                }
+            }
         }
 
+        if (agent_state == 1 || placer_opts.place_agent_multistate == false) {
+            //use the first state
+            placement_inner_loop(&state, placer_opts,
+                                 inner_recompute_limit, &stats,
+                                 &costs,
+                                 &moves_since_cost_recompute,
+                                 pin_timing_invalidator.get(),
+                                 place_delay_model.get(),
+                                 placer_criticalities.get(),
+                                 placer_setup_slacks.get(),
+                                 *move_generator,
+                                 blocks_affected,
+                                 timing_info.get(),
+                                 placer_opts.place_algorithm,
+                                 move_helper_vectors,
+                                 timing_bb_factor);
+        } else {
+            //use the second state
+            placement_inner_loop(&state, placer_opts,
+                                 inner_recompute_limit, &stats,
+                                 &costs,
+                                 &moves_since_cost_recompute,
+                                 pin_timing_invalidator.get(),
+                                 place_delay_model.get(),
+                                 placer_criticalities.get(),
+                                 placer_setup_slacks.get(),
+                                 *move_generator2,
+                                 blocks_affected,
+                                 timing_info.get(),
+                                 placer_opts.place_algorithm,
+                                 move_helper_vectors,
+                                 timing_bb_factor);
+        }
+        tot_iter += state.move_lim;
+        ++state.num_temps;
+
         print_place_status(state, stats, temperature_timer.elapsed_sec(), critical_path.delay(), sTNS, sWNS, tot_iter);
 
+        if (placer_opts.place_quench_algorithm.is_timing_driven() && agent_state == 1 && state.alpha < 0.85 && state.alpha > 0.6) {
```

Comment:
Comment why we're changing state / what this does (briefly).

Reply:
Done